### PR TITLE
catalog/lease: add new cluster setting for session based leasing

### PR DESCRIPTION
**Pull request by fqazi**
To help support session-based leasing, we will add a new internal cluster setting for tracking the leasing mode. As well as interfaces to query this setting to help add support for future session-based leasing. This setting will be used in future PRs.

Epic: CC-24173
Release note: None

### CODE REVIEW

File: pkg/sql/catalog/lease/lease.go [hunk 1]
```diff
@@ -73,6 +73,56 @@ var LeaseJitterFraction = settings.RegisterFloatSetting(
 	base.DefaultDescriptorLeaseJitterFraction,
 	settings.Fraction)
 
+//go:generate stringer -type=SessionBasedLeasingMode
```
Comment (chrisseto) [inline pkg/sql/catalog/lease/lease.go:76]: super nit: I think it's generally convention to add an extra newline here such that go:generate doesn't get picked up as the doc. string?

File: pkg/sql/catalog/lease/lease.go [hunk 2]
```diff
@@ -73,6 +73,56 @@ var LeaseJitterFraction = settings.RegisterFloatSetting(
...
+type SessionBasedLeasingMode int64
+
+const (
+	// SessionBasedLeasingOff expiry based leasing is being used.
+	SessionBasedLeasingOff SessionBasedLeasingMode = iota
+	// SessionBasedDualWrite expiry based and session based leasing are
+	// active concurrently, and both tables must be consulted schema changes.
+	SessionBasedDualWrite
+	// SessionBasedDrain expiry based leases will no longer be renewed for
```
Comment (chrisseto) [inline pkg/sql/catalog/lease/lease.go:85]: Is this saying something along the lines of "Expiry based leases will not be granted or renewed. Valid preexisting leases must still be respected."

File: pkg/sql/catalog/lease/lease.go [hunk 3]
```diff
@@ -73,6 +73,56 @@ var LeaseJitterFraction = settings.RegisterFloatSetting(
...
+	map[int64]string{
+		int64(SessionBasedLeasingOff): "off",
+		int64(SessionBasedDualWrite):  "dual_write",
+		int64(SessionBasedDrain):      "drain",
+		int64(SessionBasedOnly):       "session",
+	},
+	settings.WithReportable(false),
+)
+
+// refreshSessionBasedLeasingMode is used to refresh the active session based leasing mode.
```
Comment (chrisseto) [inline pkg/sql/catalog/lease/lease.go:110]: Manually performing the refreshes feels a bit strange to me. Could this method document when and/or why the leasing mode should be refreshed and when it's okay to have stale reads? Could this be rolled into the `get` method?

File: pkg/sql/catalog/lease/storage.go [hunk 4]
```diff
@@ -78,6 +79,11 @@ type writer interface {
 	insertLease(context.Context, *kv.Txn, leaseFields) error
 }
 
+type sessionBasedLeasingModeReader interface {
```
Comment (chrisseto) [inline pkg/sql/catalog/lease/storage.go:82]: Why isn't the refresh method exposed here?

**Review by chrisseto — APPROVED** [review]
Nothing blocking but a handful of questions and suggestions!

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111427)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto)

___
*[`pkg/sql/catalog/lease/lease.go` line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111427#gh-1365574997:-NhEv1KRBQs_owmrIMeN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8399f519dfe05bdec0a90f09f34db7679d9f756/pkg/sql/catalog/lease/lease.go#L76)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

super nit: I think it's generally convention to add an extra newline here such that go:generate doesn't get picked up as the doc. string?
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/lease/lease.go` line 85 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111427#gh-1365586633:-NhEv1d6AGyKJ1T9Zce4:b-dqwwvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8399f519dfe05bdec0a90f09f34db7679d9f756/pkg/sql/catalog/lease/lease.go#L85)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Is this saying something along the lines of "Expiry based leases will not be granted or renewed. Valid preexisting leases must still be respected."
</blockquote></details>

Yes, that's a nicer way to describe it, changed this to:

// SessionBasedDrain expiry based leases will not be granted or renewed.// Valid pre-existing leases that are expiry based will still be respected.
___
*[`pkg/sql/catalog/lease/lease.go` line 110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111427#gh-1365594377:-NhEvhO_CCEdQb_rvyWY:b-fsn4gz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8399f519dfe05bdec0a90f09f34db7679d9f756/pkg/sql/catalog/lease/lease.go#L110)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Manually performing the refreshes feels a bit strange to me. Could this method document when and/or why the leasing mode should be refreshed and when it's okay to have stale reads? Could this be rolled into the `get` method?
</blockquote></details>

Done.

Honestly, that's a good point, I"ll push this directly in the get methods. 
___
*[`pkg/sql/catalog/lease/storage.go` line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111427#gh-1365595845:-NhEwE8J3ZkSTfQnfFwR:b-hxvxkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8399f519dfe05bdec0a90f09f34db7679d9f756/pkg/sql/catalog/lease/storage.go#L82)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Why isn't the refresh method exposed here?
</blockquote></details>

Done.

Removed that method and directly using settings everywhere now.

**Comment by fqazi** [comment]
@chrisseto TFTR!

**Pull request merged**